### PR TITLE
Fix tf.experimental.numpy.isclose for integer inputs

### LINKED ISSUE #108657: tf.experimental.numpy.isclose returns incorrect results for integer inputs
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

v2.20.0-rc0-4-g72fbba3d20f 2.20.0

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

## Description:
`tf.experimental.numpy.isclose` is documented as a TensorFlow variant of NumPy's `isclose` function. It is expected to behave identically to NumPy, including support for integer inputs. However, when the input tensors are integers, `tf.experimental.numpy.isclose` return incorrect results.

## Analysis:
The expected behavior of the below reproduction code is that all elements return True, consistent with NumPy.
The discrepancy occurs only when the inputs are integer types (int32, int64).
TensorFlow does not correctly handle integer inputs internally in tf.experimental.numpy.isclose. Even though integer arithmetic should in principle work with the formula: `∣a − b∣ ≤ atol + rtol * ∣b∣`
the internal implementation appears to mishandle the comparison, resulting in all False.
**Casting inputs to float32 or float64 returns the correct results.**

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

input = tf.constant(np.array([[-13, -5, -12], [3, 17, -6], [5, 3, -8], [-18, 19, 10]]), dtype='int32')
other = tf.constant(np.array([-7]), dtype='int32')
rtol = 9
atol = 0
equal_nan = False

out_tf_int = tf.experimental.numpy.isclose(input, other, rtol=rtol, atol=atol, equal_nan=equal_nan)
print("out_tf_int:\n", out_tf_int)

out_tf_float = tf.experimental.numpy.isclose(tf.cast(input, tf.float64), tf.cast(other, tf.float64), rtol=rtol, atol=atol)
print("out_tf_float:\n", out_tf_float)

input = np.array([[-13, -5, -12], [3, 17, -6], [5, 3, -8], [-18, 19, 10]], dtype=np.int32)
other = np.array([-7], dtype=np.int32)
rtol = 9
atol = 0
equal_nan = False

out_np_int = np.isclose(input, other, rtol=rtol, atol=atol, equal_nan=equal_nan)
print("out_np_int:\n", out_np_int)

out_np_float = np.isclose(input.astype(np.float64), other.astype(np.float64), rtol=rtol, atol=atol)
print("out_np_float:\n", out_np_float)
```

### Relevant log output

```shell
out_tf_int:
 tf.Tensor(
[[False False False]
 [False False False]
 [False False False]
 [False False False]], shape=(4, 3), dtype=bool)
out_tf_float:
 tf.Tensor(
[[ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]], shape=(4, 3), dtype=bool)
out_np_int:
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]]
out_np_float:
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]]
```

### PR DESCRIPTION
**Pull request by gamila-wisam**
Fixes #108657
ready for review

Fix tf.experimental.numpy.isclose for integer inputs.

Changes include:

- Fixes incorrect results for integer inputs while keeping floating-point behavior unchanged.

- Matches NumPy semantics for integer comparisons, including support for broadcasting and tolerances.

- Adds unit tests covering integer tensors, large integers, booleans, and edge cases.

Note:

- Integer comparisons now cast to float64 only when necessary for rtol/atol.

- Exact equality (rtol=0, atol=0) is handled efficiently in integer domain.

- Large integers may still lose precision when casting to float64, consistent with NumPy behavior.

### CODE REVIEW

File: tensorflow/python/ops/numpy_ops/np_math_ops_test.py [hunk 1]
```diff
@@ -97,7 +100,7 @@ def testMinimum(self):
         np.minimum,
         'minimum',
         check_promotion_result_type=False)
-
+
```
Comment (mihaimaruseac) [inline tensorflow/python/ops/numpy_ops/np_math_ops_test.py]: No trailing whitespace, please

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/numpy_ops/np_math_ops_test.py (hunk 1 above)
Comment (gamila-wisam) [reply tensorflow/python/ops/numpy_ops/np_math_ops_test.py]: Done.
removed the trailing whitespace and pushed the fix

**Comment by gamila-wisam** [comment]
The requested change (removing trailing whitespace) has been addressed and pushed.
@mihaimaruseac PTAL

**Comment by mihaimaruseac** [comment]
No longer at Google, no longer can approve PRs

**Comment by gamila-wisam** [comment]
Thanks for letting me know.

**Comment by gamila-wisam** [comment]
Hi @cantonios ,
The requested change (removing trailing whitespace) has been addressed and pushed, Could you please review and approve when you get a chance? Thanks

File: tensorflow/python/ops/numpy_ops/np_math_ops.py [hunk 2]
```diff
@@ -534,8 +534,15 @@ def f(a, b):  # pylint: disable=missing-docstring
         result = result | (math_ops.is_nan(a) & math_ops.is_nan(b))
       return result
     else:
-      return a == b
-
+      # Numpy supports integer inputs for isclose by promoting to float
+      # we follow the same behavior to match Numpy semantics
+      a_f = math_ops.cast(a, dtypes.float64)
```
Comment (cantonios) [inline tensorflow/python/ops/numpy_ops/np_math_ops.py]: What about cases where integers don't fit in float64?

Some int64 values will end up getting truncated.  In that case, multiple _distinct_ int64 values map to the same double value, so will report here as being close - but in reality their difference could be larger than the atol.

For example, both 2^60 and (2^60+124) are both representable in int64, but both get rounded to 2^60 when converted to double.

**Comment by gamila-wisam** [comment]
Hi @cantonios , thanks for pointing this out!

well the intent here is to match NumPy’s isclose behavior by promoting integer inputs to floating point, rather than comparing integers exactly. As you mentioned, this does mean that very large int64 values (above 2^53) can lose precision when cast to float64, and distinct integers may then compare as close.

This is a known limitation of floating-point–based isclose behavior and is consistent with NumPy’s approach. The previous TensorFlow behavior (a == b) diverged from NumPy and produced incorrect results for integer inputs even in common cases.

I can add a clarifying comment or a test documenting this precision limitation if that would be helpful.

**Comment by cantonios** [comment]
> This is a known limitation of floating-point–based isclose behavior and is consistent with NumPy’s approach. The previous TensorFlow behavior (a == b) diverged from NumPy and produced incorrect results for integer inputs even in common cases.

What's an example of an incorrect behavior?  When the atol/rtol are large?

I prefer correctness over strict adherence to numpy's results.  If numpy suffers the same issue for large integers, then that's a numpy bug that should be addressed - we have encountered (and fixed) these kinds of things before in numpy.

You might try taking the difference as an integer prior to casting to double.  You'll need to check for overflows first though.
 
The other concern is performance.  Comparing integers is fast and easy on most hardware platforms.  Doubles are expensive, especially if these tensors are large.  Imagine an int8 tensor of billions of elements, then all of a sudden you cast two of them to double type.  This is also something that we care a lot more about than numpy does because of the sheer scale of some of the TF models.

I think I would err on the side of keeping it as is with the straight integer comparison.

**Comment by gamila-wisam** [comment]
Hi @cantonios, can you please take a look when you have a chance?

**Comment by keerthanakadiri** [comment]
Hi @cantonios , Can you please review this PR? Thanks!

File: tensorflow/python/ops/numpy_ops/np_math_ops.py [hunk 3]
```diff
@@ -530,11 +531,34 @@ def f(a, b):  # pylint: disable=missing-docstring
...
     else:
-      return a == b
+      # Integer dtypes (includes `bool`).
+      # Treat booleans as small integers so numeric comparisons work the
+      # same way as NumPy (e.g. True -> 1, False -> 0).
+      if dtype == dtypes.bool:
+        a = math_ops.cast(a, dtypes.int8)
+        b = math_ops.cast(b, dtypes.int8)
+
+      diff = math_ops.abs(a - b)
```
Comment (ratgr) [inline tensorflow/python/ops/numpy_ops/np_math_ops.py:547]: ∣a − b∣ ≤ atol + rtol * ∣b∣
you see in the integer world addition is commutative so
∣a - b ∣- atol  ≤  rtol * ∣b∣
so you can do the atol without loosing precision, which will handle many cases
if this goes through then you might want to submit an issue on numpy to match this.

**Review by ratgr — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/numpy_ops/np_math_ops.py (hunk 3 above)
Comment (ratgr) [reply tensorflow/python/ops/numpy_ops/np_math_ops.py:547]: you might want to do rtol * ∣b∣as doubles and cast them back to whatever dtype they originaly had so that  it is a int comparison

**Comment by ratgr** [comment]
> The other concern is performance. Comparing integers is fast and easy on most hardware platforms. Doubles are expensive, especially if these tensors are large. Imagine an int8 tensor of billions of elements, then all of a sudden you cast two of them to double type. This is also something that we care a lot more about than numpy does because of the sheer scale of some of the TF models.


I agree with [cantonios](https://github.com/cantonios) here, however this is a compatibility layer and we don't expect it to be as performant, but correctness is more important than compatibility, so try doing the fix I asked above and fix the pylint issue

**Comment by cantonios** [comment]
> > The other concern is performance. Comparing integers is fast and easy on most hardware platforms. Doubles are expensive, especially if these tensors are large. Imagine an int8 tensor of billions of elements, then all of a sudden you cast two of them to double type. This is also something that we care a lot more about than numpy does because of the sheer scale of some of the TF models.
> 
> I agree with [cantonios](https://github.com/cantonios) here, however this is a compatibility layer and we don't expect it to be as performant, but correctness is more important than compatibility, so try doing the fix I asked above and fix the pylint issue

I do care about performance here.  Casting to double is incredibly expensive and can cause a lot of large models to blow up their memory usage.  That would not be acceptable.

**Comment by ratgr** [comment]
Following internal discussion, we have determined that bypassing explicit type conversions in favor of the implementation below covers all intended use cases:

```Python
diff = tf.abs(a - b)
rhs = atol + rtol * tf.abs(b)
out = (diff <= rhs)
```

Key Considerations:
This will work if rtol and atol are integers.
Type promotion is disabled by default so this will fail on int values and float tolerances.
However if [automatic promotion](https://www.tensorflow.org/guide/tf_numpy_type_promotion) is enabled, the logic will handle promotions as expected. 
This makes bad performance an opt-in not default

This change introduces a shift in behavior; updated release notes and tests should be included to reflect these adjustments.
